crates/typst/src/math/attach.rs

```diff
@@ -465,7 +566,84 @@ fn compute_shifts_up_and_down(
+/// script's corner (tl, tr, bl, br).
+fn math_kern(
+    ctx: &MathContext,
+    base: &MathFragment,
+    script: &MathFragment,
+    shift: Abs,
+    pos: Corner,
+) -> Abs {
+    // This process is described under the MathKernInfo table in the OpenType
+    // MATH spec. Note that there is a bug in the spec (as of 2024-08-15): it
+    // says to take the minimum of the two sums, but as the kerning value is
+    // usually negative it really means the smaller kern. The current wording of
+    // the spec could result in glyphs colliding.
+    match pos {
+        // superscripts
+        Corner::TopLeft | Corner::TopRight => {
+            // Calculate two correction heights:
+            // - The distance from the base's baseline to the bottom of the
+            //   superscript's bounding box.
+            // - The distance from the superscript's baseline to the top of the
+            //   base's bounding box.
+            let corr_height_bot = shift - script.descent();
+            let corr_height_top = base.ascent() - shift;
+
+            // Calculate the sum of kerning values for each correction height.
+            let base_kern = base.kern_at_height(ctx, pos, corr_height_bot);
+            let superscript_kern =
+                script.kern_at_height(ctx, opposite_corner(pos), corr_height_bot);
+            let sum = base_kern + superscript_kern;
+
+            let base_kern = base.kern_at_height(ctx, pos, corr_height_top);
+            let superscript_kern =
+                script.kern_at_height(ctx, opposite_corner(pos), corr_height_top);
+
+            // Take the smaller kerning amount (and so the larger value).
+            sum.max(base_kern + superscript_kern)
+        }
+        // subscripts
+        Corner::BottomLeft | Corner::BottomRight => {
+            // Calculate two correction heights:
+            // - The distance from the base's baseline to the top of the
+            //   subscript's bounding box.
+            // - The distance from the subscript's baseline to the bottom of the
+            //   base's bounding box.
+            let corr_height_top = script.ascent() - shift;
+            let corr_height_bot = shift - base.descent();
+
+            // Calculate the sum of kerning values for each correction height.
+            let base_kern = base.kern_at_height(ctx, pos, corr_height_bot);
+            let subscript_kern =
+                script.kern_at_height(ctx, opposite_corner(pos), corr_height_bot);
+            let sum = base_kern + subscript_kern;
+
+            let base_kern = base.kern_at_height(ctx, pos, corr_height_top);
+            let subscript_kern =
+                script.kern_at_height(ctx, opposite_corner(pos), corr_height_top);
+
+            // Take the smaller kerning amount (and so the larger value).
+            sum.max(base_kern + subscript_kern)
+        }
+    }
+}
+
+/// Get the opposite corner.
+fn opposite_corner(position: Corner) -> Corner {
```

Comment:
I think is generally useful and could be `Corner::inv`, similar to `Dir::inv` and `Align::inv`.